engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArraySlice.java

```diff
@@ -29,26 +29,36 @@ final class ArraySlice extends EnsoObject {
   private final long start;
   private final long end;
 
-  private ArraySlice(Object storage, long start, long end) {
-    if (storage instanceof ArraySlice slice) {
+  private ArraySlice(Object base, long start, long end) {
+    var s = findStorage(base);
+    if (s instanceof ArraySlice slice) {
       this.storage = slice.storage;
       this.start = slice.start + start;
       this.end = Math.min(slice.end, slice.start + end);
     } else {
       if (CompilerDirectives.inInterpreter()) {
-        if (!InteropLibrary.getUncached().hasArrayElements(storage)) {
+        if (!InteropLibrary.getUncached().hasArrayElements(s)) {
           throw EnsoContext.get(null)
               .raiseAssertionPanic(
-                  null, "ArraySlice needs array-like delegate, but got: " + storage, null);
+                  null, "ArraySlice needs array-like delegate, but got: " + s, null);
         }
       }
-
-      this.storage = storage;
+      this.storage = s;
       this.start = start;
       this.end = end;
     }
   }
 
+  private static Object findStorage(Object storage) {
+    for (; ; ) {
+      return switch (storage) {
+        case Vector.Generic gen -> gen.toArray();
+        case ArraySlice slice -> slice;
+        default -> storage;
+      };
+    }
```

Comment:
What is the purpose of the `for` loop here if it is immediately exited via `return`?

Reply:
Looks like a relict of a previous experiment: `for` loop removed in 53aa361, thanks.